### PR TITLE
Fix #2363: better handle extractors in exhaustivity check

### LINKED ISSUE #2363: Better handling of extractors in reachability check
Dotty reports that in the following, one case is not reachable. Extractors should be better handled to avoid such annoying warnings.

```Scala
sealed trait Expr
class IntExpr extends Expr
class BooleanExpr extends Expr

object IntExpr {
  def unapply(expr: Expr): Option[IntExpr] = ???
}

object BooleanExpr {
  def unapply(expr: Expr): Option[BooleanExpr] = ???
}


class Test {
  def foo(x: List[Expr]): Int = x match {
    case IntExpr(_) :: xs => 1
    case BooleanExpr(_) :: xs => 1                              //   <==   unreachable
    case Nil => 2
  }
}
```

### PR DESCRIPTION
**Pull request by liufengyun**
Fix #2363: better handle extractors in exhaustivity check

1. Better handle exactors
2. Translate the pattern `List(pat1, pat2, c:_*)` to space semantically
3. Deal with extractors that return `Some[_]` or option-less type
4. Reduce redundant counterexamples in the warning

Previously we approximate extractor either to the extracted type space or Empty,
which is too coarse, and it causes some false positives about unreachable case.
Introducing the concept extractor into the theory enables us to inspect the
patterns inside extractors and check more unreachable cases, which is not
supported by Scalac.

Scalac is silent with following code, but Dotty _correctly_ reports both unexhaustive and
unreachable warnings.

```Scala
    sealed trait Node
    case class NodeA(i: Int) extends Node
    case class NodeB(b: Boolean) extends Node
    case class NodeC(s: String) extends Node

    object Node {
      def unapply(node: Node): Option[(Node, Node)] = ???
    }

    object Test {
      def foo(x: Node): Boolean = x match { // unexhaustive
        case Node(NodeA(_), NodeB(_)) => true
        case Node(NodeA(4), NodeB(false)) => true // unreachable code
      }
    }
```

### CODE REVIEW

File: compiler/src/dotty/tools/dotc/reporting/UniqueMessagePositions.scala [hunk 1]
```diff
@@ -23,7 +23,7 @@ trait UniqueMessagePositions extends Reporter {
         for (pos <- m.pos.start to m.pos.end) {
           positions get (ctx.source, pos) match {
             case Some(level) if level >= m.level => shouldHide = true
-            case _ => positions((ctx.source, pos)) = m.level
+            case _ => // positions((ctx.source, pos)) = m.level
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/reporting/UniqueMessagePositions.scala]: The CI fails because the change of this line. There's some bug hidden in `isHidden`, I have to comment it out to temporarily make sure all the exhaustivity/unreachability warnings are shown. Otherwise, the unreachable warning in the PR example will be incorrectly hidden.

**Comment by smarter** [comment]
@liufengyun Not directly related to this PR, but while you're working on this code, could you have a look at the exhaustiveness warnings reported against dotty itself when bootstrapping? See http://dotty-ci.epfl.ch/lampepfl/dotty/2267/4 for example. You can reproduce that locally by running `dotty-compiler-bootstrapped/compile`. It would be good to see which of these warnings are real warnings we need to fix and which one are bugs in the exhaustiveness checker.

**Comment by liufengyun** [comment]
@smarter Sure, I'll examine log to see if there're still false positives.

**Comment by liufengyun** [comment]
Now all false positives about unreachability/exhaustiveness are fixed. True positives will be addressed in https://github.com/lampepfl/dotty/issues/2429 separately.

Could you please review @smarter ?

**Comment by liufengyun** [comment]
@felixmulder The last commit 31ae57b contains some changes to the doc-tool, could you please confirm if the changes make sense?

File: doc-tool/src/dotty/tools/dottydoc/util/MemberLookup.scala [hunk 2]
```diff
@@ -77,6 +77,7 @@ trait MemberLookup {
       case (x :: xs, _) =>
         if (xs.nonEmpty) globalLookup
         else lookup(entity, packages, "scala." + query)
+      case (Nil, _) => ??? // impossible as long as query is not empty
```
Comment (felixmulder) [inline doc-tool/src/dotty/tools/dottydoc/util/MemberLookup.scala]: let's do something better than `???`, otherwise LGTM

File: doc-tool/src/dotty/tools/dottydoc/model/comment/BodyEntities.scala [hunk 3]
```diff
@@ -64,7 +64,7 @@ final case class Text(text: String) extends Inline
 abstract class EntityLink(val title: Inline) extends Inline { def link: LinkTo }
 object EntityLink {
   def apply(title: Inline, linkTo: LinkTo) = new EntityLink(title) { def link: LinkTo = linkTo }
-  def unapply(el: EntityLink): Option[(Inline, LinkTo)] = Some((el.title, el.link))
+  def unapply(el: EntityLink): Some[(Inline, LinkTo)] = Some((el.title, el.link))
```
Comment (felixmulder) [inline doc-tool/src/dotty/tools/dottydoc/model/comment/BodyEntities.scala]: Why is this needed? I would prefer not to change the type here - unapplys should be `Option`...

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 4]
```diff
@@ -426,14 +509,14 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
           case space => List(space)
         }
       case OrType(tp1, tp2) => List(Typ(tp1, true), Typ(tp2, true))
-      case _ if tp =:= ctx.definitions.BooleanType =>
+      case tp if tp =:= ctx.definitions.BooleanType =>
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: `tp.isRef(defn.BooleanClass)`

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 5]
```diff
@@ -549,6 +632,8 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
 
         if (ctx.definitions.isTupleType(tp))
           signature(tp).map(_ => "_").mkString("(", ", ", ")")
+        else if (sym.showFullName == "scala.collection.immutable.List")
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: We should never do string comparisons to determine the type of something. Instead something like `sym == defn.ListClass` should be used, if `ListClass` is not defined in `Definitions.scala` then it should be added.

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 6]
```diff
@@ -664,31 +751,35 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
...
+      // in redundancy check, take guard as false in order to soundly approximate
       val prevs = cases.take(i).map { x =>
-        if (x.guard.isEmpty) project(x.pat, false)
+        if (x.guard.isEmpty) project(x.pat)
         else Empty
       }.reduce((a, b) => Or(List(a, b)))
 
       val curr = project(cases(i).pat)
 
+      debug.println(s"---------------reachable? ${show(curr)}")
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: I suggest adding a `Printer` for the exhaustiveness checker in https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/config/Printers.scala instead of using `debug` here.

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 7]
```diff
@@ -98,19 +102,31 @@ trait SpaceLogic {
   def show(sp: Space): String
 
   /** Simplify space using the laws, there's no nested union after simplify */
-  def simplify(space: Space): Space = space match {
+  def simplify(space: Space, aggressive: Boolean = false): Space = space match {
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Could you update the documentation of this method to explain what the `aggressive` parameter is for?

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 8]
```diff
@@ -137,26 +153,33 @@ trait SpaceLogic {
...
       case (Empty, _) => true
       case (_, Empty) => false
-      case (Or(ss), _) => ss.forall(isSubspace(_, b))
+      case (Or(ss), _) =>
+        ss.forall(isSubspace(_, b))
       case (Typ(tp1, _), Typ(tp2, _)) =>
-        isSubType(tp1, tp2) || tryDecompose1(tp1) || tryDecompose2(tp2)
-      case (Typ(tp1, _), Or(ss)) =>
+        isSubType(tp1, tp2)
+      case (Typ(tp1, _), Or(ss)) =>  // optimization
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Maybe improve this comment to explain what this is an optimization of?

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 9]
```diff
@@ -386,7 +469,7 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
   /** Is `tp1` a subtype of `tp2`?  */
   def isSubType(tp1: Type, tp2: Type): Boolean = {
     // check SI-9657 and tests/patmat/gadt.scala
-    val res = erase(tp1) <:< erase(tp2)
+    val res = tp1 <:< erase(tp2)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Could you add a comment explaining why it's correct to erase `tp2` but not `tp1` ?

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 10]
```diff
@@ -420,7 +420,7 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
         Kon(pat.tpe.stripAnnots, pats.map(pat => project(pat)))
       else if (fun.symbol.owner == scalaSeqFactoryClass && fun.symbol.name == nme.unapplySeq)
         projectList(pats)
-      else if (fun.symbol.info.resultType.isRef(scalaSomeClass))
+      else if (!fun.symbol.info.finalResultType.isRef(scalaOptionClass))
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: I don't think this is correct in general, the return type of an option-less patmat may define a `def isEmpty: Boolean` method, if this method return `false`, then the case will not be taken.

Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Sorry, if `isEmpty` returns `true`, not `false`

Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Ah, I see, then it means we cannot tell from types whether an option-less is exhaustive or not. I'll retract the last commit.

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala (hunk 6 above)
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: The `debug` is just a rename of the exhaustivity printer :)

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala (hunk 10 above)
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: Well, if the method is defined as `def isEmpty: false` then we know it will always return false :). But we should probably not encode this logic in the exhaustiveness checker until we have a real spec for option-less pattern-matching that explains all cases.

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 11]
```diff
@@ -469,6 +475,19 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
   /** Is `tp1` a subtype of `tp2`?  */
   def isSubType(tp1: Type, tp2: Type): Boolean = {
     // check SI-9657 and tests/patmat/gadt.scala
+    //
+    // `erase` is a walkaround to make the following code pass the check:
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: typo: walkaround -> workaround

File: compiler/src/dotty/tools/dotc/transform/patmat/Space.scala [hunk 12]
```diff
@@ -323,11 +329,11 @@ class SpaceEngine(implicit ctx: Context) extends SpaceLogic {
   import SpaceEngine._
   import tpd._
 
-  private val scalaSomeClass       = ctx.requiredClassRef("scala.Some".toTypeName).symbol.asClass
+  private val scalaSomeClass       = ctx.requiredClass("scala.Some".toTypeName)
   private val scalaSeqFactoryClass = ctx.requiredClass("scala.collection.generic.SeqFactory".toTypeName)
   private val scalaListType        = ctx.requiredClassRef("scala.collection.immutable.List".toTypeName)
   private val scalaNilType         = ctx.requiredModuleRef("scala.collection.immutable.Nil".toTermName)
-  private val scalaConType         = ctx.requiredClassRef("scala.collection.immutable.::".toTypeName)
+  private val scalaConsType        = ctx.requiredClassRef("scala.collection.immutable.::".toTypeName)
```
Comment (smarter) [inline compiler/src/dotty/tools/dotc/transform/patmat/Space.scala]: You could get rid of all the `.toTypeName` here if `Decorators.PreNamedString` is imported.
